tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -304,6 +336,80 @@ private GraphOperation graphOp(Operand<?> operand) {
     return seen;
   }
 
+  /**
+   * Get all ops directly or indirectly required to calculate {@code outputs} (not including {@code outputs}), including
+   * control dependencies.
+   *
+   * @param outputs the starting points of the traversal.
+   * @return the ops needed to calculate {@code outputs}, not including {@code outputs}
+   */
+  public synchronized Set<GraphOperation> upstreamOps(Set<GraphOperation> outputs) {
+    Set<GraphOperation> seen = new LinkedHashSet<>(outputs.size());
+    Queue<GraphOperation> todo = new ArrayDeque<>(outputs);
+    while (!todo.isEmpty()) {
+      GraphOperation current = todo.poll();
+
+      if (seen.add(current)) {
+        todo.addAll(current.inputs().stream().map(this::graphOp).collect(Collectors.toSet()));
+        todo.addAll(current.controlInputs());
+      }
+    }
+    seen.removeAll(outputs);
+    return seen;
+  }
+
+  /**
+   * Get all ops that use one of {@code inputs} directly or indirectly (not including {@code inputs}), including control
+   * dependencies.
+   *
+   * @param inputs the starting points of the traversal.
+   * @return the ops that depend on {@code inputs}, not including {@code inputs}
+   */
+  public synchronized Set<GraphOperation> downstreamOps(Set<GraphOperation> inputs) {
+    Set<GraphOperation> seen = new LinkedHashSet<>(inputs.size());
+    Queue<GraphOperation> todo = new ArrayDeque<>(inputs);
+    while (!todo.isEmpty()) {
+      GraphOperation current = todo.poll();
+
+      if (seen.add(current)) {
+        todo.addAll(current.consumers());
+        todo.addAll(current.controlConsumers());
+      }
+    }
+    seen.removeAll(inputs);
+    return seen;
+  }
+
+  /**
+   * Get all ops directly or indirectly required to calculate {@code outputs} (not including {@code outputs}), including
+   * control dependencies.
+   *
+   * @param outputs the starting points of the traversal.
+   * @return the ops needed to calculate {@code outputs}, not including {@code outputs}
+   */
+  public synchronized Set<GraphOperation> upstream(Set<Operand<?>> outputs) {
+    return upstreamOps(outputs.stream().map(this::graphOp).collect(Collectors.toSet()));
```

Comment:
What does this do if you hand it a set of `Operand`s from `EagerOperation`s?

Reply:
`graphOp` will throw (with a nice message), since it calls `checkInput`.